delivery_postlogistics/postlogistics/web_service.py

```
@@ -0,0 +1,510 @@
+        }
+
+        # Remove the services if the delivery fixed date is not set
+        if "ZAW3217" in services:
+            if picking.delivery_fixed_date:
+                attributes["deliveryDate"] = picking.delivery_fixed_date
+            else:
+                services.remove("ZAW3217")
+
+        # parcelNo / parcelTotal cannot be used if service ZAW3218 is not activated
+        if "ZAW3218" in services:
+            if pack_total > 1:
+                attributes.update(
+                    {"parcelTotal": pack_total - 1, "parcelNo": pack_num - 1}
+                )
+            else:
+                services.remove("ZAW3218")
+
+        if "ZAW3219" in services and picking.delivery_place:
+            attributes["deliveryPlace"] = picking.delivery_place
+        if picking.carrier_id.postlogistics_proclima_logo:
+            attributes["proClima"] = True
+        else:
+            attributes["proClima"] = False
+
+        attributes["przl"] = services
+
+        return attributes
+
+    def _get_itemid(self, picking, pack_no):
+        """ Allowed characters are alphanumeric plus `+`, `-` and `_`
+        Last `+` separates picking name and package number (if any)
+
+        :return string: itemid
+
+        """
+        name = _compile_itemid.sub("", picking.name)
+        if not pack_no:
+            return name
+
+        pack_no = _compile_itemid.sub("", pack_no)
+        codes = [name, pack_no]
+        return "+".join(c for c in codes if c)
+
+    def _cash_on_delivery(self, picking, package=None):
+        if package:
+            amount = package.postlogistics_cod_amount()
+        else:
+            amount = picking.postlogistics_cod_amount()
+        amount = "{:.2f}".format(amount)
+        return [{"Type": "NN_BETRAG", "Value": amount}]
+
+    def _get_item_additional_data(self, picking, package=None):
+        result = []
+        packaging_codes = package and package.packaging_id._get_packaging_codes() or []
```

When the package has no packaging_id, we should raise an error to the user: "The package %s must have a package type"

**Traceback**
{{ packaging_codes = package and package.packaging_id._get_packaging_codes() or []}}
{{ File "/odoo/external-src/delivery-carrier/delivery_postlogistics/models/product_packaging.py", line 18, in _get_packaging_codes}}
{{ self.ensure_one()}}
{{ File "/odoo/src/odoo/models.py", line 5104, in ensure_one}}
{{ raise ValueError("Expected singleton: %s" % self)}}
ValueError: Expected singleton: product.packaging()

---

Add in https://github.com/OCA/delivery-carrier/pull/323/commits/430e52d51d740eb55e962336dd11056827aba033